src/lem/store.rs

```diff
@@ -0,0 +1,354 @@
+use rayon::prelude::*;
+use std::collections::HashMap;
+
+use dashmap::DashMap;
+use indexmap::IndexSet;
+use itertools::Itertools;
+
+use crate::{
+    field::{FWrap, LurkField},
+    hash::PoseidonCache,
+    lem::tag::Tag,
+};
+
+use super::{
+    pointers::{Ptr, ZChildren, ZPtr},
+    symbol::Symbol,
+};
+
+/// The `Store` is a crucial part of Lurk's implementation and tries to be a
+/// vesatile data structure for many parts of Lurk's data pipeline.
+///
+/// It holds Lurk data structured as trees of `Ptr`s (or `ZPtr`s). When a `Ptr`
+/// has children`, we store them in the `IndexSet`s available: `ptrs2`, `ptrs3`
+/// or `ptrs4`. These data structures speed up LEM interpretation because lookups
+/// by indices are fast.
+///
+/// The `Store` also provides an infra to speed up interning strings and symbols.
+/// This data is saved in `vec_char_cache` and `vec_str_cache`, which are better
+/// explained in `intern_string` and `intern_symbol_path` respectively.
+///
+/// There's also a process that we call "hydration", in which we use Poseidon
+/// hashes to compute the sstable) hash of the children of a pointer. These hashes
+/// are necessary when we want to create Lurk proofs because the circuit consumes
+/// elements of the `LurkField`, not (unstable) indices of `IndexSet`s.
+///
+/// Lastly, we have a `HashMap` to hold commited data, which can be retrieved by
+/// the resulting commitment hash.
+#[derive(Default)]
+pub struct Store<F: LurkField> {
+    ptrs2: IndexSet<(Ptr<F>, Ptr<F>)>,
+    ptrs3: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>)>,
+    ptrs4: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, Ptr<F>)>,
+
+    vec_char_cache: HashMap<Vec<char>, Ptr<F>>,
```

Comment:
I don't see a good reason why this should be `Vec<char>` instead of `String`. Is there one? I think using `String` will save space and is less confusing if there's not a strong reason to avoid.

Reply:
It's better explained in `intern_string`. The vector is used because the cache is stored in reverse order. And due to the mechanics of that function, using the vector directly is the most performant approach because we avoid unnecessary conversions between `Vec<char>` and `String`